### PR TITLE
enable torch autocast for mps [PRI-341]

### PR DESCRIPTION
**Pull request by jmkuebler**
## Issue

Partially adresses PriorLabs/TabPFN#1102

## Motivation and Context

We currently disable autocast on MPS. However, enabling it gives drastic speeupds on Apple Silikon. We are not aware of any accuracy concerns specific to MPS.

By now it seems that bfloat16 is supported by torch for MPS, since `torch==2.6.0` [https://github.com/pytorch/pytorch/pull/139390#issuecomment-2666900556](<https://github.com/pytorch/pytorch/pull/139390#issuecomment-2666900556>).

It reduces the runtime of below code snippet from around 8.3s to 5s on my macbook

```python
from tabpfn import TabPFNClassifier, TabPFNRegressor
from sklearn.datasets import make_classification, make_regression
import torch

from time import time

X_train, y_train = make_classification(
            n_samples=1000,
            n_features=100,
            n_classes=2,
        )

X_test = X_train

start = time()
clf = TabPFNClassifier()
clf.fit(X_train, y_train)  # downloads checkpoint on first use
predictions = clf.predict(X_test)
print(predictions[0])
print("time to fit and predict", time()-start)
```

---

## Public API Changes

- [ ] No Public API changes
- [X] Yes, Public API changes (Details below)

---

## Checklist

- [X] The changes have been tested locally.
- [ ] Documentation has been updated (if the public API or usage changes).
- [ ] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
- [X] The code follows the project's style guidelines.
- [X] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

**Comment by jmkuebler** [comment]
On a fresh python 3.12 venv with torch 2.5.0 (so before bfloat support for autocast on MPS was merged) I get on my branch:
```
==== 128 failed, 1340 passed, 152 skipped, 8 xfailed, 2333 warnings in 335.58s (0:05:35) ====
```

on main we get:

```
============== 128 failed, 1340 passed, 152 skipped, 8 xfailed, 2333 warnings in 313.26s (0:05:13) ===============
```

So those python/torch versions are anyways not an option and should'nt hold us back.

**Review by klemens-floege — COMMENTED** [review]
@jmkuebler the PR looks great, you mentioned you get failing tests for: 

```
requires-python = ">=3.10"
"torch>=2.5",
```

When looking our toml these do seem to be supported, but are not tested on CI. Shall we hold off on merging and ask if we want to bump to torch 2.6 or so? :)

**Comment by jmkuebler** [comment]
I would support bumping torch to 2.6, but that rather comes from the fact that bfloat 16 is only part of autocast for MPS since then.

WRT to the test, the same number of tests fail on main. So it is not related to my change. Yet I'd propose to see if there are objections to bump to 2.6

**Comment by jmkuebler** [comment]
holding back until https://github.com/PriorLabs/TabPFN/pull/1126 lands
Reveritng to draft until then

**Comment by jmkuebler** [comment]
ok #1126 is approved and in the merge queue. Could you then approve this one @klemens-floege ?

**Comment by klemens-floege** [comment]
I think there is still some lack of clarity whether we want to bump to 2.6

**Comment by jmkuebler** [comment]
We merged https://github.com/PriorLabs/TabPFN/pull/1129 which checks that torch is >=2.6 for MPS. This should enable our PR.
@klemens-floege

**Review by klemens-floege — APPROVED** [review]
Amazing, thanks a lot!
